tests/functional/test_ec2.py

```diff
@@ -0,0 +1,66 @@
+# Copyright 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from tests import unittest
+from botocore.stub import Stubber, StubAssertionError
+import botocore.session
+
+
+class TestIdempotencyToken(unittest.TestCase):
+    def setUp(self):
+        self.region = 'us-west-2'
+        self.session = botocore.session.get_session()
+        self.client = self.session.create_client(
+            'ec2', self.region)
+        self.stubber = Stubber(self.client)
+
+    def test_insert_idempotency_token(self):
+        service_response = {}
+        expected_params_notoken = {
+            'PurchaseRequests': [
+                {'PurchaseToken': 'foo',
+                 'InstanceCount': 123}],
+        }
+        expected_params_token = {
+            'PurchaseRequests': [
+                {'PurchaseToken': 'foo',
+                 'InstanceCount': 123}],
+            'ClientToken': 'foobar'
+        }
+        function_name = 'purchase_scheduled_instances'
+
+        # First one should inject a token, second one is passed one
+        self.stubber.add_response(
+            function_name, service_response, expected_params_notoken)
+        self.stubber.add_response(
+            function_name, service_response, expected_params_token)
+
+        self.stubber.activate()
+
+        # Check the ClientToken parameter exsits
+        # after injection and before the call
+        self.client.meta.events.register_first(
```

Comment:
1) Sounds good
2) I tried that first and it threw fit because I think the token gets inserted after the stubber looks at it? I'm not sure I can try that way again and fiddle with it more. I just figured the stubber didn't support what I wanted to do so I went around it.
3) Will do.